File: test/language/sql_tables/java/sql_tables/blob_param_table/BlobParamTableTest.java
package sql_tables.blob_param_table;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.io.File;
import java.io.IOException;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

import org.junit.After;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;

import test_utils.FileUtil;
import test_utils.JdbcUtil;

import sql_tables.TestDb;

import zserio.runtime.ZserioError;
import zserio.runtime.array.UnsignedIntArray;

public class BlobParamTableTest
{
    @BeforeClass
    public static void init()
    {
        JdbcUtil.registerJdbc();
    }

    @Before
    public void setUp() throws IOException, SQLException
    {
        FileUtil.deleteFileIfExists(file);
        database = new TestDb(file.toString());
        database.createSchema();
    }

    @After
    public void tearDown() throws SQLException
    {
        if (database != null)
        {
            database.close();
            database = null;
        }
    }

    @Test
    public void deleteTable() throws SQLException
    {
        assertTrue(isTableInDb());

        final BlobParamTable testTable = database.getBlobParamTable();
        testTable.deleteTable();
        assertFalse(isTableInDb());

        testTable.createTable();
        assertTrue(isTableInDb());
    }

    @Test
    public void readWithoutCondition() throws SQLException, IOException, ZserioError
    {
        final BlobParamTable testTable = database.getBlobParamTable();

        final List<BlobParamTableRow> writtenRows = new ArrayList<BlobParamTableRow>();
        fillBlobParamTableRows(writtenRows);
        testTable.write(writtenRows);

        final List<BlobParamTableRow> readRows = testTable.read();
        checkBlobParamTableRows(writtenRows, readRows);
    }

    @Test
    public void readWithCondition() throws SQLException, IOException, ZserioError
    {
        final BlobParamTable testTable = database.getBlobParamTable();

        final List<BlobParamTableRow> writtenRows = new ArrayList<BlobParamTableRow>();
        fillBlobParamTableRows(writtenRows);
        testTable.write(writtenRows);

        final String condition = "name='Name1'";
        final List<BlobParamTableRow> readRows = testTable.read(condition);

        assertEquals(1, readRows.size());

        final int expectedRowNum = 1;
        final BlobParamTableRow readRow = readRows.get(0);
        checkBlobParamTableRow(writtenRows.get(expectedRowNum), readRow);
    }

    @Test
    public void update() throws SQLException, IOException, ZserioError
    {
        final BlobParamTable testTable = database.getBlobParamTable();

        final List<BlobParamTableRow> writtenRows = new ArrayList<BlobParamTableRow>();
        fillBlobParamTableRows(writtenRows);
        testTable.write(writtenRows);

        final int updateRowId = 3;
        final BlobParamTableRow updateRow = createBlobParamTableRow(updateRowId, "UpdatedName");
        final String updateCondition = "blobId=" + updateRowId;
        testTable.update(updateRow, updateCondition);

        final List<BlobParamTableRow> readRows = testTable.read(updateCondition);
        assertEquals(1, readRows.size());

        final BlobParamTableRow readRow = readRows.get(0);
        checkBlobParamTableRow(updateRow, readRow);
    }

    @Test
    public void nullValues() throws SQLException, IOException, ZserioError
    {
        final BlobParamTable testTable = database.getBlobParamTable();

        final List<BlobParamTableRow> writtenRows = new ArrayList<BlobParamTableRow>();
        fillBlobParamTableRowsWithNullValues(writtenRows);
        testTable.write(writtenRows);

        final List<BlobParamTableRow> readRows = testTable.read();
        checkBlobParamTableRows(writtenRows, readRows);
    }

    private static void fillBlobParamTableRows(List<BlobParamTableRow> rows)
    {
        for (int blobId = 0; blobId < NUM_BLOB_PARAM_TABLE_ROWS; ++blobId)
            rows.add(createBlobParamTableRow(blobId, "Name" + blobId));
    }

    private static BlobParamTableRow createBlobParamTableRow(int blobId, String name)
    {
        final BlobParamTableRow row = new BlobParamTableRow();
        row.setBlobId(blobId);
        row.setName(name);
        final Parameters parameters = new Parameters(PARAMETERS_COUNT);
        row.setParameters(parameters);
        final UnsignedIntArray array = new UnsignedIntArray(PARAMETERS_COUNT);
        for (int i = 0; i < PARAMETERS_COUNT; ++i)
            array.setElementAt(i, i);
        final ParameterizedBlob parameterizedBlob = new ParameterizedBlob(parameters, array);
        row.setBlob(parameterizedBlob);

        return row;
    }

    private static void fillBlobParamTableRowsWithNullValues(List<BlobParamTableRow> rows)
    {
        for (int blobId = 0; blobId < NUM_BLOB_PARAM_TABLE_ROWS; ++blobId)
            rows.add(createBlobParamTableRowWithNullValues(blobId));
    }

    private static BlobParamTableRow createBlobParamTableRowWithNullValues(int blobId)
    {
        final BlobParamTableRow row = new BlobParamTableRow();
        row.setBlobId(blobId);
        row.setName(null);
        row.setParameters(null);
        row.setBlob(null);

        return row;
    }

    private static void checkBlobParamTableRows(List<BlobParamTableRow> rows1,
            List<BlobParamTableRow> rows2)
    {
        assertEquals(rows1.size(), rows2.size());
        for (int i = 0; i < rows1.size(); ++i)
            checkBlobParamTableRow(rows1.get(i), rows2.get(i));
    }

    private static void checkBlobParamTableRow(BlobParamTableRow row1, BlobParamTableRow row2)
    {
        assertEquals(row1.getBlobId(), row2.getBlobId());
        if (row1.isNullName())
            assertTrue(row2.isNullName());
        else
            assertEquals(row1.getName(), row2.getName());
        if (row1.isNullParameters())
            assertTrue(row2.isNullParameters());
        else
            assertEquals(row1.getParameters(), row2.getParameters());
        if (row1.isNullBlob())
            assertTrue(row2.isNullParameters());
        else
            assertEquals(row1.getBlob(), row2.getBlob());
    }

    private boolean isTableInDb() throws SQLException
    {
        // check if database does contain table
        final String sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + TABLE_NAME +
                "'";

        final PreparedStatement statement = database.connection().prepareStatement(sqlQuery);
        try
        {
            final ResultSet resultSet = statement.executeQuery();
            if (!resultSet.next())
                return false;

            // read table name
            final String tableName = resultSet.getString(1);
            if (resultSet.wasNull() || !tableName.equals(TABLE_NAME))
                return false;
        }
        finally
        {
            statement.close();
        }

        return true;
    }

    private static final String TABLE_NAME = "blobParamTable";

    private static final int    PARAMETERS_COUNT = 10;
    private static final int    NUM_BLOB_PARAM_TABLE_ROWS = 5;
    private static final String FILE_NAME = "blob_param_table_test.sqlite";

    private final File file = new File(FILE_NAME);
    private TestDb database = null;
}

File: test/extensions/language/sql_tables/python/BlobParamTableTest.py
import os

import SqlTables


class BlobParamTableTest(SqlTables.TestCaseWithDb):
    def setUp(self):
        if os.path.exists(self.dbFileName):
            os.remove(self.dbFileName)
        self._database = self.api.TestDb.from_file(self.dbFileName)
        self._database.create_schema()

    def tearDown(self):
        self._database.close()

    def testDeleteTable(self):
        self.assertTrue(self._isTableInDb())

        testTable = self._database.blob_param_table
        testTable.delete_table()
        self.assertFalse(self._isTableInDb())

        testTable.create_table()
        self.assertTrue(self._isTableInDb())

    def testWriteColumns(self):
        testTable = self._database.blob_param_table

        writtenRows = []
        writtenRows.append(self._createBlobParamTableRow(999, "Pepek"))
        writtenRows.append(self._createBlobParamTableRow(999, "Rishi"))

        columns = ["name"]
        testTable.write(writtenRows, columns)

        readRows = testTable.read()
        for i, readRow in enumerate(readRows):
            self.assertEqual(i + 1, readRow[testTable.Column.BLOB_ID])
            self.assertEqual(writtenRows[i][testTable.Column.NAME], readRow[testTable.Column.NAME])
            self.assertTrue(readRow[testTable.Column.BLOB] is None)
            self.assertTrue(readRow[testTable.Column.PARAMETERS] is None)

    def testReadWithoutCondition(self):
        testTable = self._database.blob_param_table

        writtenRows = self._createBlobParamTableRows()
        testTable.write(writtenRows)

        readRows = testTable.read()
        numReadRows = 0
        for readRow in readRows:
            self.assertEqual(writtenRows[numReadRows], readRow)
            numReadRows += 1

        self.assertTrue(len(writtenRows), numReadRows)

    def testReadWithCondition(self):
        testTable = self._database.blob_param_table

        writtenRows = self._createBlobParamTableRows()
        testTable.write(writtenRows)

        condition = "name='Name1'"
        readRows = testTable.read(condition)
        expectedRowNum = 1
        for readRow in readRows:
            self.assertEqual(writtenRows[expectedRowNum], readRow)

    def testReadColumns(self):
        testTable = self._database.blob_param_table

        writtenRows = self._createBlobParamTableRows()
        testTable.write(writtenRows)

        condition = "name='Name1'"
        columns = ["blobId", "name"]
        readRows = testTable.read_columns(columns, condition)
        expectedRowNum = 1
        for readRow in readRows:
            self.assertEqual(
                writtenRows[expectedRowNum][testTable.Column.BLOB_ID], readRow[testTable.Column.BLOB_ID]
            )
            self.assertEqual(writtenRows[expectedRowNum][testTable.Column.NAME], readRow[testTable.Column.NAME])
            self.assertTrue(readRow[testTable.Column.BLOB] is None)
            self.assertTrue(readRow[testTable.Column.PARAMETERS] is None)

    def testUpdate(self):
        testTable = self._database.blob_param_table

        writtenRows = self._createBlobParamTableRows()
        testTable.write(writtenRows)

        updateRowId = 3
        updateRow = self._createBlobParamTableRow(updateRowId, "UpdatedName")
        updateCondition = "blobId=" + str(updateRowId)
        testTable.update(updateRow, updateCondition)

        readRows = testTable.read(updateCondition)
        for readRow in readRows:
            self.assertEqual(updateRow, readRow)

    def testUpdateColumns(self):
        testTable = self._database.blob_param_table

        writtenRows = self._createBlobParamTableRows()
        testTable.write(writtenRows)

        updateRowId = 3
        columns = ["name"]
        updateRow = self._createBlobParamTableRow(-999, "Pepek")
        updateCondition = "blobId=" + str(updateRowId)
        testTable.update_columns(updateRow, columns, updateCondition)

        writtenRow = writtenRows[updateRowId]
        self.assertEqual(writtenRow[testTable.Column.BLOB_ID], updateRowId)
        readRows = testTable.read(updateCondition)
        for readRow in readRows:
            for i, readColumn in enumerate(readRow):
                if i == testTable.Column.NAME:
                    self.assertEqual(updateRow[i], readColumn)
                else:
                    self.assertEqual(writtenRow[i], readColumn)

    def testNullValues(self):
        testTable = self._database.blob_param_table

        writtenRows = self._createBlobParamTableRowsWithNullValues()
        testTable.write(writtenRows)

        readRows = testTable.read()
        numReadRows = 0
        for readRow in readRows:
            self.assertEqual(writtenRows[numReadRows], readRow)
            numReadRows += 1

        self.assertTrue(len(writtenRows), numReadRows)

    def _createBlobParamTableRows(self):
        rows = []
        for blobId in range(self.NUM_BLOB_PARAM_TABLE_ROWS):
            rows.append(self._createBlobParamTableRow(blobId, "Name" + str(blobId)))

        return rows

    def _createBlobParamTableRow(self, blobId, name):
        parameters = self.api.blob_param_table.Parameters(self.PARAMETERS_COUNT)
        array = list(range(self.PARAMETERS_COUNT))
        blob = self.api.blob_param_table.ParameterizedBlob(parameters, array)

        return (blobId, name, parameters, blob)

    def _createBlobParamTableRowsWithNullValues(self):
        rows = []
        for blobId in range(self.NUM_BLOB_PARAM_TABLE_ROWS):
            rows.append(self._createBlobParamTableRowWithNullValues(blobId))

        return rows

    @staticmethod
    def _createBlobParamTableRowWithNullValues(blobId):
        return (blobId, None, None, None)

    def _isTableInDb(self):
        # check if database does contain table
        sqlQuery = "SELECT name FROM sqlite_master WHERE type='table' AND name='" + self.TABLE_NAME + "'"
        for row in self._database.connection.cursor().execute(sqlQuery):
            if len(row) == 1 and row[0] == self.TABLE_NAME:
                return True

        return False

    TABLE_NAME = "blobParamTable"

    PARAMETERS_COUNT = 10
    NUM_BLOB_PARAM_TABLE_ROWS = 5
